File: packages/typescript/src/models/workflow.ts
import type { YAMLMap } from "yaml";
import type { HttpsJsonSchemastoreOrgGithubWorkflowJson as SchemaWorkflow } from "../schema/workflow-types.generated.js";
import { buildModel, extractMeta } from "./_base.js";
import type {
  OnModel,
  PermissionsModel,
  ConcurrencyModel,
  DefaultsModel,
  JobModel,
  WithMeta,
  Raw,
  ModelSpec,
  WorkflowModel,
} from "./_base.js";
import type { OnInput } from "./trigger.js";
import { on } from "./trigger.js";
import type { PermissionsInput } from "./permissions.js";
import { permissions } from "./permissions.js";
import type { ConcurrencyInput } from "./job.js";
import { concurrency, defaults } from "./job.js";
import type { DefaultsInput } from "./job.js";

/**
 * Input for the top-level workflow model, representing a complete GitHub
 * Actions workflow YAML file. Contains the workflow name, triggers,
 * permissions, environment variables, defaults, concurrency settings,
 * and jobs.
 */
export interface WorkflowInput {
  /** The display name of the workflow. */
  name?: string;
  /** Custom name for workflow runs. Supports GitHub Actions expressions. Serialized as `run-name`. */
  runName?: string;
  /** Trigger configuration for the workflow. Accepts a pre-built `OnModel` or an inline `OnInput` object which will be auto-wrapped. */
  on?: OnModel | OnInput;
  /** Token permissions. Can be a `PermissionsModel`, an inline `PermissionsInput`, a string shorthand (`"read-all"` / `"write-all"`), or a `Raw<string>` for arbitrary values. */
  permissions?: PermissionsModel | PermissionsInput | "read-all" | "write-all" | Raw<string>;
  /** Environment variables available to all jobs in the workflow. */
  env?: Record<string, string>;
  /** Default settings for all `run` steps. Accepts a pre-built `DefaultsModel` or an inline `DefaultsInput`. */
  defaults?: DefaultsModel | DefaultsInput;
  /** Concurrency group configuration. Can be a string (group name), a `ConcurrencyModel`, or an inline `ConcurrencyInput`. */
  concurrency?: string | ConcurrencyModel | ConcurrencyInput;
  /** Map of job IDs to job definitions. At least one job is required. Values can be `JobModel` objects or raw `YAMLMap` nodes for passthrough. */
  jobs: Record<string, JobModel | YAMLMap>;
}

/** Serialization spec for {@link WorkflowModel}. */
export const WORKFLOW_SPEC: ModelSpec = {
  kind: "workflow",
  fieldMap: {
    name: "name",
    runName: "run-name",
    on: "on",
    permissions: "permissions",
    env: "env",
    defaults: "defaults",
    concurrency: "concurrency",
    jobs: "jobs",
  } satisfies Record<keyof WorkflowInput, keyof SchemaWorkflow>,
  order: ["name", "run-name", "on", "permissions", "env", "defaults", "concurrency", "jobs"],
  wrap: {
    on: { factory: on, mode: "model" },
    permissions: { factory: permissions, mode: "objectModel" },
    defaults: { factory: defaults, mode: "objectModel" },
    concurrency: { factory: concurrency, mode: "objectModel" },
  },
};

/**
 * Create a workflow model representing a complete GitHub Actions workflow
 * YAML file. Plain-object values for `on`, `permissions`, `defaults`, and
 * `concurrency` are automatically wrapped with their respective factory
 * functions.
 *
 * @param input - Workflow properties and optional model metadata.
 * @returns A `WorkflowModel` that can be serialized to YAML.
 *
 * @example
 * ```ts
 * workflow({
 *   name: "CI",
 *   on: { push: { branches: ["main"] } },
 *   jobs: {
 *     test: job({
 *       runsOn: "ubuntu-latest",
 *       steps: [step({ uses: "actions/checkout@v4" })],
 *     }),
 *   },
 * })
 * ```
 */
export function workflow(input: WithMeta<WorkflowInput>): WorkflowModel {
  const [data, meta] = extractMeta(input);
  return buildModel<WorkflowModel>(WORKFLOW_SPEC, data as Record<string, unknown>, meta);
}

File: packages/python/src/ghagen/models/workflow.py
"""Top-level Workflow model."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import Field

from ghagen._raw import Raw
from ghagen.emitter.key_order import WORKFLOW_KEY_ORDER
from ghagen.models._base import Document, OrRaw
from ghagen.models.job import Concurrency, Defaults, Job
from ghagen.models.permissions import Permissions
from ghagen.models.trigger import On


class Workflow(Document):
    """A GitHub Actions workflow definition.

    This is the top-level model that maps to a complete workflow YAML file.

    Example::

        workflow = Workflow(
            name="CI",
            on=On(push=PushTrigger(branches=["main"])),
            jobs={
                "test": Job(
                    runs_on="ubuntu-latest",
                    steps=[Step(uses="actions/checkout@v4")],
                ),
            },
        )
        print(workflow.to_yaml())
    """

    name: str | None = None
    run_name: str | None = Field(
        None,
        serialization_alias="run-name",
        description="Custom name for workflow runs, supports expressions.",
    )
    on: OrRaw[On | dict[str, Any]] | None = None
    permissions: (
        OrRaw[Permissions | Literal["read-all", "write-all"] | Raw[str]] | None
    ) = None
    env: dict[str, str] | None = None
    defaults: OrRaw[Defaults] | None = None
    concurrency: OrRaw[str | Concurrency] | None = None
    jobs: dict[str, OrRaw[Job]] = Field(default_factory=dict)

    def _get_key_order(self) -> list[str]:
        return WORKFLOW_KEY_ORDER
